PGO: Add new tiers

EgorBo opened the pull request:
This PR implements @jkotas's idea in https://github.com/dotnet/runtime/issues/70410#issuecomment-1149874663 when `DOTNET_TieredPGO` is enabled (it's off by default and will be so in .NET 7.0)

> 1) Use R2R code for process startup
> 2) Once the process startups up, switch to instrumented code for hot methods
> 3) Once you collect enough PGO data, create optimized JITed version

# Design

```mermaid
flowchart
    prestub(.NET Function) -->|Compilation| hasAO{"Marked with<br/>[AggressiveOpts]?"}
    hasAO-->|Yes|tier1ao["JIT to <b><ins>Tier1</ins></b><br/><br/>(no dynamic profile data)"]
    hasAO-->|No|hasR2R
    hasR2R{"Is prejitted (R2R)?"} -->|No| tier000

    tier000["JIT to <b><ins>Tier0</ins></b><br/><br/>(not optimized, not instrumented,<br/> with patchpoints)"]-->|Running...|ishot555
    ishot555{"Is hot?<br/>(called >30 times)"}
    ishot555-.->|No,<br/>keep running...|ishot555
    ishot555-->|Yes|tier0
   
    hasR2R -->|Yes| R2R
    R2R["Use <b><ins>R2R</ins></b> code<br/><br/>(optimized, not instrumented,<br/>no patchpoints)"] -->|Running...|ishot1
    ishot1{"Is hot?<br/>(called >30 times)"}-.->|No,<br/>keep running...|ishot1
    ishot1--->|"Yes"|tier1inst

    tier0["JIT to <b><ins>Tier0Instrumented</ins></b><br/><br/>(not optimized, instrumented,<br/> with patchpoints)"]-->|Running...|ishot5
    tier1pgo2["JIT to <b><ins>Tier1</ins></b><br/><br/>(optimized with profile data)"]
      
    tier1inst["JIT to <b><ins>Tier1Instrumented</ins></b><br/><br/>(optimized, instrumented, <br/>no patchpoints)"]
    tier1inst-->|Running...|ishot5
    ishot5{"Is hot?<br/>(called >30 times)"}-->|Yes|tier1pgo2
    ishot5-.->|No,<br/>keep running...|ishot5

```

jkotas commented:
cc @noahfalk @kouvel

EgorBo commented:
Current results:
![image](https://user-images.githubusercontent.com/523221/174455917-b11d13e2-5105-40cb-b65f-a0b39238a181.png)

Both `base` and `diff` use a locally built coreclr that matches SDK (latest daily) and both use `DOTNET_TieredPGO=1`. `base` results match the official results - OK. `diff` is much slower for some reason but the fact that it compiled more functions makes perfect sense. I am downloading traces for both cases now to analyze jit events

EgorBo commented:
Ah looks like my approach to check whether we compile from R2R is wrong
```
Assert failure(PID 31732 [0x00007bf4], Thread: 43456 [0xa9c0]): !MethodDescBackpatchInfoTracker::IsLockOwnedByCurrentThread() || IsInForbidSuspendForDebuggerRegion()

CORECLR! Thread::RareDisablePreemptiveGC + 0x131 (0x00007ff8`deafb481)
CORECLR! Thread::DisablePreemptiveGC + 0xC5 (0x00007ff8`de562d75)
CORECLR! GCCoopHackNoThread::GCCoopHackNoThread + 0x113 (0x00007ff8`de6c2393)
CORECLR! HashMap::LookupValue + 0x172 (0x00007ff8`de6c37a2)
CORECLR! ReadyToRunInfo::GetMethodDescForEntryPointInNativeImage + 0x1CD (0x00007ff8`de86fa1d)
CORECLR! ReadyToRunInfo::GetMethodDescForEntryPoint + 0xBB (0x00007ff8`de86f7fb)
CORECLR! TieredCompilationManager::AsyncPromoteToTier1 + 0x22D (0x00007ff8`de81f88d)
```

EgorBo commented:
So far I don't see any errors in the logs but it seems in a highly concurrent environment (I managed to create a small repro) it stops promoting such methods to tier1 and I don't understand why, keep looking..

EgorBo commented:
Looks like the problem that I have to reset callcounting cell and its state since new one is not created

EgorBo commented:
Question: say, method A() calls B() which is R2R'd - as far as I understand A() is going to call it through a callcounting intermediate method that looks like this:

![image](https://user-images.githubusercontent.com/523221/174501921-1c75fde7-ff0a-4a14-8dde-a27bc24dfb62.png)

what happens with this method when B() is promoted to tier1? is its body erased and replaced with direct call to tier1 version of B()? or only TargetForMethod and we still hit the call counting cell each call?

janvorli commented:
> 2\. "Patch" the initial [callcounting] stub to look at tier0 instead of r2r and reset the call-counting-cell

It sounds better to me, as we never delete the call counting stubs, so this would leave less garbage.

noahfalk commented:
When I was doing tiered compilation originally one of the benchmarks I found helpful was [MusicStore](https://github.com/aspnet/MusicStore) and I had the app measure its own latency for requests 0-500, 501-1000, 1000-1500, and so on. This helped me get an idea how quickly an app was able to converge to the steady state behavior. Completely up to you if a similar analysis would be useful now. One hypothesis I'd have for the worse TE numbers is that the benchmark might be short enough that it is capturing a substantial amount of pre-steady-state behavior.

EgorBo commented:
> When I was doing tiered compilation originally one of the benchmarks I found helpful was [MusicStore](https://github.com/aspnet/MusicStore) and I had the app measure its own latency for requests 0-500, 501-1000, 1000-1500, and so on. This helped me get an idea how quickly an app was able to converge to the steady state behavior. Completely up to you if a similar analysis would be useful now. One hypothesis I'd have for the worse TE numbers is that the benchmark might be short enough that it is capturing a substantial amount of pre-steady-state behavior.

Thanks! When I will be patching the call-counting stub we might consider resetting the call-counting-cell to some smaller number (e.g. 10)

EgorBo commented:
For FullPGO mode where we don't have R2R my prototype shows +10% improvement for "Startup time + time to first request", something like this:
![image](https://user-images.githubusercontent.com/523221/174914602-893dd944-e4f3-471b-9ab8-37b9eb3667ab.png)

(it's `tier0` -> 30 calls -> `tier0 with instrumentation` -> 30 calls -> `tier1`) - worth trying different threshold in-between

EgorBo commented:
## `DOTNET_TieredPGO=1`
```
| application           | eeeg_plaintext_mvc_Pgo_0_base | eeeg_plaintext_mvc_Pgo_0_diff |        |
| --------------------- | ----------------------------- | ----------------------------- | ------ |
| CPU Usage (%)         |                            99 |                           100 | +1.01% |
| Cores usage (%)       |                         2,764 |                         2,799 | +1.27% |
| Working Set (MB)      |                           577 |                           580 | +0.52% |
| Private Memory (MB)   |                         1,770 |                         1,770 |  0.00% |
| Build Time (ms)       |                         3,708 |                         3,741 | +0.89% |
| Start Time (ms)       |                           267 |                           268 | +0.37% |
| Published Size (KB)   |                       111,441 |                       111,441 |  0.00% |
| .NET Core SDK Version |     7.0.100-preview.6.22316.8 |     7.0.100-preview.6.22316.8 |        |
| load                   | eeeg_plaintext_mvc_Pgo_0_base | eeeg_plaintext_mvc_Pgo_0_diff |         |
| ---------------------- | ----------------------------- | ----------------------------- | ------- |
| CPU Usage (%)          |                            30 |                            34 | +13.33% |
| Cores usage (%)        |                           827 |                           941 | +13.78% |
| Working Set (MB)       |                            38 |                            38 |   0.00% |
| Private Memory (MB)    |                           358 |                           358 |   0.00% |
| Start Time (ms)        |                             0 |                             0 |         |
| First Request (ms)     |                           136 |                           136 |   0.00% |
| Requests/sec           |                     2,263,203 |                     2,892,576 | +27.81% |
| Requests               |                    34,173,291 |                    43,675,784 | +27.81% |
| Mean latency (ms)      |                          1.14 |                          0.92 | -19.30% |
| Max latency (ms)       |                         81.62 |                         50.21 | -38.48% |
| Bad responses          |                             0 |                             0 |         |
| Socket errors          |                             0 |                             0 |         |
| Read throughput (MB/s) |                        284.90 |                        364.13 | +27.81% |
| Latency 50th (ms)      |                          1.05 |                          0.80 | -23.52% |
| Latency 75th (ms)      |                          1.53 |                          1.18 | -22.88% |
| Latency 90th (ms)      |                          2.01 |                          1.58 | -21.39% |
| Latency 99th (ms)      |                          0.00 |                          0.00 |         |
```
+25-28% RPS (and Latency) improvement without regressions for `First Request (ms)` 

## `DOTNET_TieredPGO=1` and `DOTNET_ReadyToRun=0` (aka FullPGO mode)
```
| application           | eeeg_plaintext_mvc_FullPgo_1_base | eeeg_plaintext_mvc_FullPgo_1_diff |        |
| --------------------- | --------------------------------- | --------------------------------- | ------ |
| CPU Usage (%)         |                                96 |                                98 | +2.08% |
| Cores usage (%)       |                             2,699 |                             2,758 | +2.19% |
| Working Set (MB)      |                               568 |                               572 | +0.70% |
| Private Memory (MB)   |                             1,766 |                             1,770 | +0.23% |
| Build Time (ms)       |                             3,748 |                             3,769 | +0.56% |
| Start Time (ms)       |                               770 |                               701 | -8.96% |
| Published Size (KB)   |                           111,441 |                           111,441 |  0.00% |
| .NET Core SDK Version |         7.0.100-preview.6.22316.8 |         7.0.100-preview.6.22316.8 |        |
| load                   | eeeg_plaintext_mvc_FullPgo_1_base | eeeg_plaintext_mvc_FullPgo_1_diff |         |
| ---------------------- | --------------------------------- | --------------------------------- | ------- |
| CPU Usage (%)          |                                34 |                                34 |   0.00% |
| Cores usage (%)        |                               966 |                               952 |  -1.45% |
| Working Set (MB)       |                                38 |                                38 |   0.00% |
| Private Memory (MB)    |                               358 |                               358 |   0.00% |
| Start Time (ms)        |                                 0 |                                 0 |         |
| First Request (ms)     |                               309 |                               285 |  -7.77% |
| Requests/sec           |                         2,987,104 |                         2,949,697 |  -1.25% |
| Requests               |                        45,104,462 |                        44,539,799 |  -1.25% |
| Mean latency (ms)      |                              0.89 |                              0.90 |  +1.12% |
| Max latency (ms)       |                             55.03 |                             44.60 | -18.95% |
| Bad responses          |                                 0 |                                 0 |         |
| Socket errors          |                                 0 |                                 0 |         |
| Read throughput (MB/s) |                            376.03 |                            371.32 |  -1.25% |
| Latency 50th (ms)      |                              0.82 |                              0.80 |  -2.44% |
| Latency 75th (ms)      |                              1.20 |                              1.17 |  -2.50% |
| Latency 90th (ms)      |                              1.77 |                              1.59 | -10.17% |
| Latency 99th (ms)      |                              0.00 |                              0.00 |         |
```
For Full PGO `-9%` improvement for `Start Time (ms)` and `-8%` for `First Request (ms)` because we instrument only hot tier0 methods 🙂

AndyAyersMS commented:
We can't currently leave Tier1-OSR and get back to Tier0 (at least mid-method; if the method is called every so often we could switch at a call boundary. But it wouldn't help say `Main` which is only called once).

What we could do instead is go from Tier0 (uninstrumented) to Tier0-OSR (instrumented) and then to Tier1-OSR (instrumented). This would give some PGO data, but we might not see the early parts of the method execute with instrumentation.

EgorBo commented:
I assume what you suggest looks like this:

```mermaid
graph TD
    prestub(.NET Function) --> hasAO{"Marked with<br/>[AggressiveOpts]?"}
    hasAO-->|Yes|tier1ao[JIT to Tier1]
    hasAO-->|No|hasR2R
    hasR2R{"Is AOT'd <br/> (R2R)"?} -->|No| tier0[JIT to Tier0]
    tier0 --> osr
    hasR2R -->|Yes| osr{Is OSR<br /> requested?}
    osr -->|Yes, e.g. hot loop<br/>in the method| tier0osr[JIT to Tier0-OSR<br/>with instrumentation] 
    tier0osr --> osr2{Is OSR<br/>requested}
    osr2 -->|No, but the whole<br/>method is invoked<br/>>30 times|pgoenabled
    osr2 -->|Yes, e.g. hot loop<br/>in the method| tier1osr[JIT to Tier1-OSR<br/>] 
    osr -->|No, but the whole<br/>method is invoked<br/>>30 times| pgoenabled
    tier1osr-->|method is invoked<br/>>30 times| pgoenabled{"Is DOTNET_TieredPGO=1?"}
    pgoenabled-->|Yes|tier0inst[JIT to Tier0 with <br/>instrumentation]
    pgoenabled --> |No|tier1[JIT to Tier1]
    tier0inst --> |method is invoked<br/>>30 times|tier1pgo[JIT to Tier1<br/>using profile data]
```

EgorBo committed:
R2R -> InstrumentedTier0 for hot R2R methods

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into tier0instrumented

EgorBo commented:
Ok, so I decided to limit this PR to only "Hot R2R to Instrumented Tier0" and leave "Tier0 to Instrumented Tier0" for a separate PR as it significantly complicates flow (mainly due to OSR) and test matrix, for some methods it might end up with 5 different versions per single method (e.g. `"tier0" -> "tier0 OSR instrumented" -> "tier1 OSR" -> "tier0 instrumented" -> "tier1"`).

So this PR still maintains improvements shown in https://github.com/dotnet/runtime/pull/70941#issuecomment-1163069395 for `TieredPGO=1` (~25% more RPS for PlaintexMVC)

The current flow looks like this:
```mermaid
graph TD
    prestub(.NET Function) -->|Compilation| hasAO{"Marked with<br/>[AggressiveOpts]?"}
    hasAO-->|Yes|tier1ao["<b><ins>Tier1</ins></b><br/><br/>(typically used where<br/> we can't afford tier0)"]
    hasAO-->|No|hasR2R
    hasR2R{"Is AOT'd <br/> (R2R)"?} -->|No| tier0
    hasR2R -->|Yes| R2R
    R2R["Use <b><ins>R2R</ins></b> code<br/><br/>(optimized code,<br/>no instrumentation,<br/>no patchpoints)"] -->|Running...|ishot1
    ishot1{"Is hot?<br/>(called >30 times)"}-->|No|R2R
    ishot1-->|Yes|tier0inst
    tier0inst[<b><ins>Tier0</ins></b><br/>with instrumentation] --> |Running...|ishot2
    ishot2{"Is hot?<br/>(called >30 times)"} -->|Yes| tier1pgo
    ishot2-->|No|tier0inst
    tier1pgo[<b><ins>Tier1</ins></b> optimized with profile data]
    
    tier0[<b><ins>Tier0</ins></b> with<br/>instrumentation and<br/>patchpoints for OSR] -->|Running...| osr
    osr{Did a patchpoint<br/>trigger OSR?} -->|Yes| tier1osr
    osr -->|No| ishot3
    ishot3{"Is hot?<br/>(called >30 times)"} -->|No| tier0
    ishot3 -->|Yes|tier1pgo
    tier1osr["<b><ins>Tier1-OSR</ins></b> with profile data<br/><br/>(method is partially Tier0 with<br/>instrumentation mixed with<br/>Tier1 optimized with PGO)"] -->|Running...|ishot4
    ishot4{"Is hot?<br/>(called >30 times)"}-->|Yes|tier1pgo
    ishot4-->|No|tier1osr
```

Here is a test program in two variations:
1) It's precompiled to R2R and executed: https://gist.github.com/EgorBo/b9b3d55939f6ce92c45d03d12dbe8a3a
2) Not precompiled: https://gist.github.com/EgorBo/f786f23dcbf93023922022a009aa34be

I think it's ready for review once it passes tests

EgorBo commented:
/azp list

EgorBo commented:
/azp run runtime-coreclr pgo, runtime-coreclr libraries-pgo

EgorBo commented:
PTAL @AndyAyersMS @janvorli @noahfalk @kouvel

EgorBo committed:
Oops, it turns out that I didn't properly patch re-used callcountingstub

EgorBo committed:
remove TieredPGO

EgorBo committed:
fix bug in the importer

EgorBo committed:
Add ability to instrument optimized code

EgorBo commented on src/coreclr/inc/clrconfigvalues.h:
```diff
@@ -572,11 +572,16 @@ RETAIL_CONFIG_DWORD_INFO(INTERNAL_HillClimbing_GainExponent,
 #define TC_CallCountThreshold (2)
 #define TC_CallCountingDelayMs (1)
 #define TC_DelaySingleProcMultiplier (2)
+#define TC_DelayPgoDivider (1)
```
NOTE: this divider is needed for `R2R -> tier0inst -> tier1` case because once we switch to tier0 that doesn't inline code - we start to raise a lot of "first-time" compilations each of which delays promotion for the pendingTier1 methods in the background (by 100ms). However, when we do `R2R -> tier1inst ->tier1` (with `DOTNET_TC_InstrumentOptimizedCode=1` - we're fine even with the existing number.

EgorBo committed:
fix 32bit

EgorBo committed:
fix ret type

EgorBo commented:
@AndyAyersMS good news!! `R2R -> Tier1-instr -> Tier1` works well, at least for the Plaintext MVC - same RPS improvement (+25%) 🎉 

And since the instrumentation tier is now able to inline methods it does not produce new compilations any more (as you were worried about)👍

EgorBo committed:
Clean up

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into tier0instrumented

EgorBo commented:
/azp run runtime-coreclr pgo, runtime-coreclr libraries-pgo

EgorBo commented:
6% more RPS for YARP proxy benchmark `TieredPGO=1(Main) vs TieredPGO=1(PR)`
Running all of them now

EgorBo committed:
Clean up

EgorBo commented:
/azp run runtime-coreclr pgo, runtime-coreclr libraries-pgo

EgorBo committed:
test fix

EgorBo commented:
/azp run runtime-coreclr pgo, runtime-coreclr libraries-pgo

EgorBo committed:
test2

EgorBo commented:
Any feedback? The PR is ready for the case when instrumented tier is not optimized and needs a few touches for the optimizated one (around tail calls + block counters) but I'd love to hear some feedback first.

So this PR improves only DOTNET_TieredPGO=1 mode (not enabled by default in Main) - see benchmark results in the first comment, however, In theory we could enable DOTNET_TieredPGO=1 for only R2R'd methods in .NET 7.0 - it e.g. will give us +8% more RPS for YARP while maintaining the same start up speed. In that case we need to focus on testing.

AndyAyersMS commented on src/coreclr/vm/codeversion.h:
```diff
@@ -74,12 +74,14 @@ class NativeCodeVersion
     enum OptimizationTier
     {
         OptimizationTier0,
+        OptimizationTier0Instrumented,
```
You may need to make changes similar to #66507, https://github.com/dotnet/diagnostics/pull/2928 and https://github.com/microsoft/perfview/pull/1584 to ensure this new tier and existing tieres still get reported properly by diagnostic tools.

AndyAyersMS commented on src/coreclr/vm/tieredcompilation.cpp:
```diff
@@ -1018,7 +1043,20 @@ CORJIT_FLAGS TieredCompilationManager::GetJitFlags(PrepareCodeConfig *config)
+                    flags.Set(CORJIT_FLAGS::CORJIT_FLAG_BBINSTR);
+                    if (CLRConfig::GetConfigValue(CLRConfig::UNSUPPORTED_TC_OptimizedInstrumentedTier) != 0)
+                    {
+                        flags.Set(CORJIT_FLAGS::CORJIT_FLAG_TIER1);
```
I think this decision should be driven by a different native code version. We'd like to be able to determine in SOS/ETW what kind of code was produced (in particular, if the method was optimized).

EgorBo replied:
Done

AndyAyersMS commented on src/coreclr/vm/tieredcompilation.cpp:
```diff
@@ -1041,6 +1079,22 @@ CORJIT_FLAGS TieredCompilationManager::GetJitFlags(PrepareCodeConfig *config)
+                flags.Set(CORJIT_FLAGS::CORJIT_FLAG_BBINSTR);
+                if (CLRConfig::GetConfigValue(CLRConfig::UNSUPPORTED_TC_OptimizedInstrumentedTier) != 0)
+                {
+                    flags.Set(CORJIT_FLAGS::CORJIT_FLAG_TIER1);
```
Similar comment as above.

EgorBo replied:
Done

AndyAyersMS commented on src/coreclr/vm/eeconfig.cpp:
```diff
@@ -740,6 +740,16 @@ HRESULT EEConfig::sync()
+            // tieredCompilation_CallCountingDelayMs
+            tieredCompilation_CallCountingDelayMs /= 3;
+            tieredCompilation_CallCountingDelayMs = max(1, tieredCompilation_CallCountingDelayMs);
+        }
```
This changes the default behavior (eg current "full pgo") too, right?

EgorBo replied:
Fixed to don't change default behavior

AndyAyersMS reviewed (commented):
It would be helpful to me to more clearly describe which version you think we should enable by default, and with this, how each of default/dynamic-pgo/full-pgo will change with this PR.

Also you should get somebody on the VM side (preferably @kouvel) to review that part more carefully.

noahfalk commented:
Sorry I was out sick. @kouvel is definitely the one we'd want to review this but he is out on vacation until next week.

Nice perf results! : )

davidwrighton commented on src/coreclr/vm/codeversion.h:
```diff
@@ -74,12 +74,14 @@ class NativeCodeVersion
     enum OptimizationTier
     {
         OptimizationTier0,
+        OptimizationTier0Instrumented,
```
As this enum is converted to an integer and directly emitted to the eventtrace stream as part of the jit information, we cannot change the values of existing enum entries such as OptimizationTier1. You need to put this enum value at the end of the OptimizationTier enum. Also, please add a comment about this to the source code.

EgorBo replied:
Done

davidwrighton reviewed (changes requested):
Keeping existing diagnostic tools working is really important, and we cannot afford to break the wire format they depend on.

@kouvel needs to review the VM side of things. In addition though, as we're adding these various variables and controls, I think we need a Book of the Runtime document in docs\design\coreclr\botr which describes the flow behavior of the various flags. Then, as we adjust the rules, we can review changes to that document, as well as the actual product code, which is much harder to read. I think we could start with a document which is basically just the flowchart you've put here in the PR comments, but we should eventually grow the thing to describe how all of the various flags interact. The document also needs to specify the default behavior that we ship with, and which knobs customers are expected to use.

I'm not sure where the tipping point in terms of complexity was passed, but we're clearly at the point of we need a descriptive document.

From what I can see here though, I really like the behavior change for TieredPGO=1, but I feel that we should not change the default for TieredPGO to 1 at this point in the release. We just don't have much runway to fix issues we may encounter, and this PR is showing that there are lots of little details that need to be understood and finalized we move everything forward. We can be a bit more cavalier about these things for non-default behaviors, as we expect that consumers of non-default options are a little more familiar with their quirks, but changes to default behavior need to be of extremely high quality.

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into new-tier

# Conflicts:
#	src/coreclr/inc/clrconfigvalues.h
#	src/coreclr/jit/fginline.cpp

EgorBo committed:
Address feedback

EgorBo committed:
Don't re-use callcountingstubs for now

EgorBo committed:
Clean up

EgorBo commented on src/coreclr/jit/fginline.cpp:
```diff
@@ -808,7 +809,10 @@ Compiler::fgWalkResult Compiler::fgLateDevirtualization(GenTree** pTree, fgWalkD
+            if (comp->fgComputePredsDone)
+            {
+                comp->fgRemoveRefPred(bNotTaken, block);
+            }
```
I reverted my JIT changes and made instrumented tier unoptimized by default for now

EgorBo commented on src/coreclr/inc/clrconfigvalues.h:
```diff
@@ -617,7 +618,15 @@ CONFIG_DWORD_INFO(INTERNAL_OSR_HighId, W("OSR_HighId"), 10000000, "High end of e
-RETAIL_CONFIG_DWORD_INFO(INTERNAL_TieredPGO, W("TieredPGO"), 0, "Instrument Tier0 code and make counts available to Tier1")
+
+// Enabled for CI tests
+RETAIL_CONFIG_DWORD_INFO(INTERNAL_TieredPGO, W("TieredPGO"), 1, "Instrument Tier0 code and make counts available to Tier1")
```
Yeah it was a CI test, reverted

EgorBo commented on src/coreclr/vm/callcounting.cpp:
```diff
@@ -618,6 +652,19 @@ bool CallCountingManager::SetCodeEntryPoint(
 
             do
             {
+                if (callCountingStage == CallCountingInfo::Stage::PendingReset)
```
Reverted to always create new callcountingstubs for this PR

EgorBo committed:
Address feedback

EgorBo commented on src/coreclr/vm/prestub.cpp:
```diff
@@ -1205,6 +1205,7 @@ PrepareCodeConfig::JitOptimizationTier PrepareCodeConfig::GetJitOptimizationTier
             switch (config->GetCodeVersion().GetOptimizationTier())
             {
                 case NativeCodeVersion::OptimizationTier0:
+                case NativeCodeVersion::OptimizationTier0Instrumented:
```
Done

EgorBo committed:
Address feedback

EgorBo commented on src/coreclr/vm/codeversion.cpp:
```diff
@@ -333,6 +333,13 @@ NativeCodeVersion::OptimizationTier NativeCodeVersion::GetOptimizationTier() con
     }
 }
 
+bool NativeCodeVersion::IsUnoptimizedTier() const
```
Renamed to `IsFinalTier`

EgorBo commented on src/coreclr/vm/tieredcompilation.cpp:
```diff
@@ -1018,7 +1043,20 @@ CORJIT_FLAGS TieredCompilationManager::GetJitFlags(PrepareCodeConfig *config)
             if (g_pConfig->TieredCompilation_QuickJit())
             {
-                _ASSERTE(nativeCodeVersion.GetOptimizationTier() == NativeCodeVersion::OptimizationTier0);
+                if (nativeCodeVersion.GetOptimizationTier() == NativeCodeVersion::OptimizationTier::OptimizationTier0Instrumented)
```
Changed code

EgorBo committed:
Add docs

EgorBo commented:
@kouvel @davidwrighton @AndyAyersMS 

I think I've addressed all of your concerns & feedback
1) I made the behavior completely opt-in at this point (so when user only enabled TieredPGO=1 they will also need to change `DOTNET_TieredPGO_Strategy` to observe the new behaviour.
2) InstrumentationTier is unoptimized by default - I'll work separately on debugging small jit issues in InstrumentationTierOptimized (around tail calls)
3) I'll file quick PRs similar to https://github.com/dotnet/diagnostics/pull/2928 and https://github.com/microsoft/perfview/pull/1584 to those repos to add new tiers there as well if you approve the current PR.
4) I added docs
5) I don't patch existing call-counting-stubs to avoid race conditions for now, will investigate how much memory they consume on close-to-real-world projects

kouvel commented:
> I don't patch existing call-counting-stubs to avoid race conditions for now, will investigate how much memory they consume on close-to-real-world projects but I assume not much - they're pretty small

They can add up to a small but not-insignificant amount. Call counting stubs/infos used to be deleted, at the moment in .NET 7 they are not deleted. I think it would be reasonable to use different call counting stubs and make them deletable again (perhaps leaking much less memory than currently).

kouvel commented:
I won't be able to fully review this soon as I'll be on vacation, I'll take a closer look when I get back. cc @mangod9

EgorBo committed:
Enable other strategies

EgorBo committed:
Clean up

EgorBo committed:
Fix GetInitialOptimizationTier

EgorBo committed:
Enable optimized Instrumentations

EgorBo committed:
Update diagram

EgorBo committed:
Fix assert

EgorBo committed:
Update

EgorBo committed:
Add test

EgorBo committed:
Update clrconfigvalues.h

CI Test

EgorBo commented:
/azp run runtime-coreclr pgo, runtime-coreclr libraries-pgo

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into new-tier

EgorBo committed:
Update clrconfigvalues.h

EgorBo commented:
/azp run runtime-coreclr pgo, runtime-coreclr libraries-pgo

EgorBo committed:
Update clrconfigvalues.h

kouvel commented on src/coreclr/vm/prestub.cpp:
```diff
@@ -1302,10 +1310,7 @@ bool PrepareCodeConfig::FinalizeOptimizationTierForTier0LoadOrJit()
         // (Tier 0 to Optimized), such that the tier is not changed in an unexpected way or at an unexpected time. Since changes
         // to the optimization tier are unlocked, this assertion is just a speculative check on possible values.
-        NativeCodeVersion::OptimizationTier previousOptimizationTier = GetCodeVersion().GetOptimizationTier();
-        _ASSERTE(
-            previousOptimizationTier == NativeCodeVersion::OptimizationTier0 ||
-            previousOptimizationTier == NativeCodeVersion::OptimizationTierOptimized);
+        _ASSERTE(!GetCodeVersion().IsFinalTier());
```
I think the previous assertion is more correct. The optimization tier could already be `OptimizationTierOptimized` due to the race mentioned in the comment above. The optimization tier here may be any initial tier or `OptimizationTierOptimized`, so after this change it may also be `OptimizationTierInstrumented`.

kouvel commented on src/coreclr/vm/tieredcompilation.cpp:
```diff
@@ -1016,9 +1088,24 @@ CORJIT_FLAGS TieredCompilationManager::GetJitFlags(PrepareCodeConfig *config)
         if (!methodDesc->RequestedAggressiveOptimization())
         {
+            NativeCodeVersion::OptimizationTier currentTier = nativeCodeVersion.GetOptimizationTier();
+            if (currentTier == NativeCodeVersion::OptimizationTier::OptimizationTierInstrumented)
+            {
+                flags.Set(CORJIT_FLAGS::CORJIT_FLAG_BBINSTR);
+                flags.Set(CORJIT_FLAGS::CORJIT_FLAG_TIER0);
+                return flags;
+            }
```
If `QuickJit` is disabled, it looks like this would still generate instrumented tier 0 code. Would it make sense to move this block inside the `if (g_pConfig->TieredCompilation_QuickJit())` block below?

kouvel commented on src/coreclr/vm/tieredcompilation.cpp:
```diff
@@ -1041,6 +1128,16 @@ CORJIT_FLAGS TieredCompilationManager::GetJitFlags(PrepareCodeConfig *config)
     {
+        case NativeCodeVersion::OptimizationTierInstrumented:
+            flags.Set(CORJIT_FLAGS::CORJIT_FLAG_BBINSTR);
+            flags.Set(CORJIT_FLAGS::CORJIT_FLAG_TIER0);
```
If `QuickJit` is disabled, this may still generate instrumented tier 0 code. I figure this could happen with an `R2R -> OptimizationTierInstrumented` transition. If `QuickJit` is disabled, is that transition disallowed somehow?

Could add an assertion that `QuickJit` is enabled here to verify that this path is not reached when it is disabled.

EgorBo committed:
Merge branch 'new-tier' of github.com:EgorBo/runtime-1 into new-tier

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into new-tier

# Conflicts:
#	src/coreclr/vm/eeconfig.cpp
#	src/coreclr/vm/prestub.cpp

EgorBo committed:
Resolve conflicts, address feedback

EgorBo committed:
Apply suggestions from code review

Co-authored-by: Koundinya Veluri <kouvel@users.noreply.github.com>

EgorBo committed:
Remove InstrumentedTiers_strat4.csproj

EgorBo committed:
Address feedback

EgorBo committed:
Merge branch 'new-tier' of github.com:EgorBo/runtime-1 into new-tier

EgorBo commented on src/coreclr/vm/eeconfig.h:
```diff
@@ -51,6 +51,28 @@ enum ParseCtl {
+    // In these modes we never instrument Tier0 and only promote hot Tier0 and R2R
+    // code to intermediate tiers with instrumentation
+    PromoteHotTier0ToInstrumentedTier = 3,
+    PromoteHotTier0ToInstrumentedTierOptimized = 4,
```
Thanks! Makes perfect sense, I've removed this strategy

EgorBo committed:
update DynamicPgo-InstrumentedTiers.md

EgorBo committed:
Clean up

EgorBo committed:
Clean up

EgorBo committed:
clean up

EgorBo committed:
Rename arg in AsyncPromoteToTier1

EgorBo committed:
Address feedback

EgorBo committed:
Address feedback

EgorBo commented on src/coreclr/vm/tieredcompilation.cpp:
```diff
@@ -1016,9 +1088,24 @@ CORJIT_FLAGS TieredCompilationManager::GetJitFlags(PrepareCodeConfig *config)
+            if (currentTier == NativeCodeVersion::OptimizationTier::OptimizationTierInstrumentedOptimized)
+            {
+                flags.Set(CORJIT_FLAGS::CORJIT_FLAG_BBINSTR);
+                flags.Set(CORJIT_FLAGS::CORJIT_FLAG_TIER1);
```
After I addressed all of your feedback it now sends TIER1 flag to instrumented tier after R2R (only if a corresponding strategy was selected)

kouvel replied:
I meant whether it would make sense to not send the tier 1 flag for even the optimized instrumented tier. It would still be optimized without that flag. Wondering whether the tier 1 flag may in the future enable more optimizations that use extra CPU time, which may not be worthwhile for an instrumented tier.

EgorBo replied:
Ok, going to replace FLAG_TIER1 with FLAG_OPT here, for now both will have the same effects but at least we won't be printing "Tier1" for JitDisasm for instrumented tiers in the comments

EgorBo committed:
Address feedback

EgorBo commented:
/azp run runtime-coreclr pgo, runtime-coreclr libraries-pgo

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into new-tier

EgorBo committed:
Fix issues found during testing

EgorBo commented:
/azp run runtime-coreclr pgo, runtime-coreclr libraries-pgo

EgorBo commented:
@kouvel thanks for the detailed feedback - think I've addressed everything, can you take a look again?

kouvel reviewed (approved):
LGTM, thanks!

AndyAyersMS commented on src/coreclr/jit/importer.cpp:
```diff
@@ -9602,7 +9602,16 @@ var_types Compiler::impImportCall(OPCODE                  opcode,
+            call = gtNewLconNode((__int64)opts.jitFlags->GetRawFlags());
+            goto DONE_CALL;
+        }
+#endif
```
Should we only recognize this "intrinsic" under some config setting? Or else make it into a real / official intrinsic?

EgorBo replied:
I'll prepare a public api proposal

jkotas replied:
Can we use tracing for this instead of intrinsics like this?

We need to be able to replay the sequence of tiered compilations, without instrumenting user code.

EgorBo replied:
@jkotas I needed this intrinsic to make sure the correct version of code was applied and **is executed**, it helped me to catch a bug: everything looked fine on the traces level, JitDisasm also displayed codegen for all code versions but it turns out one of them was not applied because I forgot to update EntryPoint (it was in the initial impl where I was patching existing call counting stubs instead of creating new ones) - I can delete this one now

jkotas replied:
You should be able to figure out using tracing what kind of code is executing.

EgorBo replied:
Makes sense, I don't really need this now so will remove once the outerloop tests finish

AndyAyersMS reviewed (commented):
* I can't tell from this what happens to OSR methods when PGO is enabled. Did the default change? Seems like OSR+PGO methods should always be considered hot and so always get instrumented.
* `COMPlus_WritePGOData` no longer enables instrumentation. Seems like that's ok, likely nobody is depending on this
* You have a few asm diffs in the asp.net collection

gbalykov commented:
@EgorBo comments in https://github.com/dotnet/runtime/issues/70410 are already closed, so I'll comment here. For `full` pgo mode you can try to use MultiCoreJit to improve startup. Profile can be collected without any R2R images or with all R2R images (i.e. first launch will still be fast), and startup during second launch will be improved too.

cc @alpencolt

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into new-tier

# Conflicts:
#	src/coreclr/jit/importer.cpp

EgorBo committed:
Simplify PR, get rid of strategies

EgorBo commented:
I've simplified this PR to have just a single strategy that makes sense for us: promote hot tier0 to unoptimized instrumentation tier and promote hot r2r to optimized instrumentation tier. I've updated the diagram in the main description to reflect that ^.

I tested this strategy on two apps - clieant app AvaloniaILSpy and a huge web app BingSNR

For [AvaloniaILSpy](https://github.com/icsharpcode/AvaloniaILSpy) startup (start, render a window with content and quit in OnLoaded handler):
1) Default mode (PublishReadyToRun): 1.61s
2) TieredPGO=1: 1.71s (most of the code is prejitted so PGO is not applied)
3) TieredPGO=1 + this PR: 1.61s - **no overhead from DynamicPGO!**
4) R2R=0: 2.03s
5) R2R=0, TieredPGO=1: 2.26s
6) R2R=0, TieredPGO=1 + this PR: 2.03s - **no overhead from DynamicPGO!**

For BingSNR:
- I'm still measuring impact but can already detect noticeable improvements around throughput and latency for Bing because it instruments R2R'd code
- Number of "methods jitted" is increased from 205k to 245k via this PR but no effect on working set size (according to my calculations it was expected to be +0.2% (or +15Mb) to the working set (which is 7.5Gb) but I don't even see it because of fluctuations).

EgorBo commented:
@AndyAyersMS Current behavior around OSR:

1) R2R:
- R2R'd version doesn't need OSR as it's fast
- R2R is promoted to InstrumentedTierOptimized, no OSR (it's optimized already)
- InstrumentedTierOptimized is promoted to Tier1

2) ILOnly
- ILOnly is compiled to Tier0 without opts, with OSR patchpoints
- Potentially promoted to OSR, still no instrumentation
- The whole method is hot so is promoted to InstrumentedTier, OSR'd version is dropped
- OSR kicks in again and optimized the loop but still instrumented
- The whole method is promoted to Tier1

Potential improvements for ILOnly path:
1) If OSR is triggered for non-instrumented method it might start instrumentation
2) if previous version of method was not instrumented but made it to OSR we might want to use "InstrumentedTierOptimized" for it (and avoid future OSR).

Number of methods with OSR in BingSNR: 60 among 245.000

EgorBo committed:
Enable TieredPGO_InstrumentOnlyHotCode by default

AndyAyersMS commented:
I like this direction -- seems like there's not much point instrumenting OSR until we know the method can be called a lot, otherwise it will never tier up anyways.

AndyAyersMS commented:
> Number of methods with OSR in BingSNR: 60 among 245.000

Also note OSR thresholds were set pretty cautiously. It's possible we should be transitioning sooner, but I don't have much data to go on here.

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into new-tier

EgorBo committed:
Fix an assert in optimized instrumented tier

EgorBo commented:
/azp run runtime-coreclr pgo, runtime-coreclr libraries-pgo

EgorBo committed:
fix osr tests

EgorBo committed:
Update docs

EgorBo committed:
Update DynamicPgo-InstrumentedTiers.md

EgorBo commented:
Updated the docs, covered OSR, working set impact, performance, start time, etc.

EgorBo commented:
/azp run runtime-coreclr pgo, runtime-coreclr libraries-pgo

EgorBo commented:
@kouvel could you please take a look one more time at the VM side? Since the previous part your approved I simplified it a lot by getting rid of all strategies except just one we want to eventually focus one.

kouvel commented on src/coreclr/vm/tieredcompilation.cpp:
```diff
@@ -261,10 +277,39 @@ void TieredCompilationManager::AsyncPromoteToTier1(
+#ifdef FEATURE_PGO
+    if (g_pConfig->TieredPGO())
+    {
+        if (currentNativeCodeVersion.IsDefaultVersion() &&
```
Probably could omit the check for the default code version. I think the profiler APIs could be used to add a new IL code version and then there could be a tier 0 native code version that is not the default code version, which probably should be tiered up the same way.

kouvel reviewed (approved):
Just one comment, otherwise the tiering stuff LGTM, thanks!

EgorBo committed:
Address feedback

EgorBo commented:
/azp run runtime-coreclr pgo, runtime-coreclr libraries-pgo

EgorBo commented:
The problem with the queue of methods pending the call counting installation (not even promotion to tier1, just to start counting calls):

![image](https://user-images.githubusercontent.com/523221/196762808-fa8d5fef-6f53-496e-ba50-a0c31f3093d9.png)

X axis - time in seconds after service launch.
Y axis - number of methods waiting for getting a call-counting stub

EgorBo commented:
Anyway, I'd like to work on improvements for that queue in a separate PR (and will adjust the OSR limit) 

@AndyAyersMS can you review/approve the jit part?
@davidwrighton this PR is blocked on your change request, could you take a look if I addressed your concerns (lack of docs)?

AndyAyersMS commented on src/coreclr/jit/compiler.cpp:
```diff
@@ -4095,7 +4095,7 @@ const char* Compiler::compGetTieringName(bool wantShortName) const
     else if (tier1)
     {
-        if (opts.jitFlags->IsSet(JitFlags::JIT_FLAG_OSR))
+        if (opts.IsOSR())
```
Seems like we should extend the logic in this method to describe the new "tier1 / instrumented case"

AndyAyersMS commented on src/coreclr/jit/fgprofile.cpp:
```diff
@@ -1887,8 +1887,8 @@ PhaseStatus Compiler::fgPrepareToInstrumentMethod()
-    const bool osrMethod            = opts.IsOSR();
-    const bool useEdgeProfiles = (JitConfig.JitEdgeProfiling() > 0) && !prejit && !tier0WithPatchpoints && !osrMethod;
+    const bool instrOpt             = opts.IsInstrumentedOptimized();
+    const bool useEdgeProfiles = (JitConfig.JitEdgeProfiling() > 0) && !prejit && !tier0WithPatchpoints && !instrOpt;
```
Do we really need block profiles for full instrumented/optimized methods? Seems like edge profiles might work -- unless perhaps the think you need is the is special handling for tail calls.

if so can you add a clarifying comment here?

EgorBo replied:
Unfortunately we still need it for now, I'll allocate some time to investigate what is needed to enable edge-profiling after this PR

AndyAyersMS commented on src/coreclr/jit/importercalls.cpp:
```diff
@@ -1310,14 +1310,16 @@ var_types Compiler::impImportCall(OPCODE                  opcode,
             // Only schedule importation if we're not currently importing.
             //
-            if (mustImportEntryBlock && (compCurBB != fgEntryBB))
+            if ((opts.IsInstrumentedOptimized() || opts.IsOSR()) && mustImportEntryBlock && (compCurBB != entryBb))
```
This is an OSR only issue, as full method compilation will naturally start importing with `fgFirstBB`.

AndyAyersMS commented on src/coreclr/inc/dacprivate.h:
```diff
@@ -610,6 +610,8 @@ struct MSLAYOUT DacpTieredVersionData
         OptimizationTier_OptimizedTier1,
         OptimizationTier_ReadyToRun,
         OptimizationTier_OptimizedTier1OSR,
+        OptimizationTier_InstrumentedTier,
```
I kind of think this should be `InstrumentedTier0` (though we don't use Tier0 here -- so arguably `InstrumentedQuickJitted`). Otherwise, I have to keep reminding myself that this is not optimized code.

Likewise below I'd use `InstrumentedTier1Optimized`.

EgorBo replied:
Good idea, renamed to match existing conventions

AndyAyersMS commented on src/coreclr/jit/importercalls.cpp:
```diff
@@ -1288,7 +1288,7 @@ var_types Compiler::impImportCall(OPCODE                  opcode,
         // * We only instrument root method blocks in OSR methods,
         //
-        if (opts.IsOSR() && !compIsForInlining())
+        if ((opts.IsInstrumentedOptimized() || opts.IsOSR()) && !compIsForInlining())
```
As noted below, I think this is an OSR only problem.

AndyAyersMS replied:
So it could be 
```suggestion
        if ((opts.IsInstrumentedOptimized() && opts.IsOSR()) && !compIsForInlining())
```

EgorBo replied:
`opts.IsInstrumentedOptimized() && opts.IsOSR()` is not working, for some reason it hits an assert even in non PGO OSR mode. I should be able to revert these changes once I fix "edge profiling" for optimized code

AndyAyersMS reviewed (commented):
Overall looks good. Left notes on a few things to consider.

I take it `COMPlus_WritePGOData` still works as long as you also enable TieredPGO? It is occasionally quite useful.

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into new-tier

EgorBo committed:
Address Andy's feedback

EgorBo committed:
Disable edge-profiling with a comment

EgorBo committed:
fix assert

EgorBo commented:
/azp run runtime-coreclr pgo, runtime-coreclr libraries-pgo

AndyAyersMS commented on docs/design/features/DynamicPgo-InstrumentedTiers.md:
````diff
@@ -236,36 +229,33 @@ G_M43040_IG08:              ;; offset=0062H
 ; Total bytes of code 116
 ```
-The codegen looks a bit bulky but if we look closer we'll see that we clonned the loop to have a fast version with a devirtualized call inside (see `G_M43040_IG03`) without guards inside. To summarize what happened with `HotLoop` we can take a look at this part of the diagram:
+The codegen looks a bit bulky but if we look closer we'll see that we clonned the loop to have a fast version with a devirtualized call inside (see `G_M43040_IG03`) with guards hoisted out of that loop. To summarize what happened with `HotLoop` we can take a look at this part of the diagram:
````
```suggestion
The codegen looks a bit bulky but if we look closer we'll see that we cloned the loop to have a fast version with a devirtualized call inside (see `G_M43040_IG03`) with guards hoisted out of that loop. To summarize what happened with `HotLoop` we can take a look at this part of the diagram:
```

EgorBo committed:
Update docs/design/features/DynamicPgo-InstrumentedTiers.md

Co-authored-by: Andy Ayers <andya@microsoft.com>

EgorBo committed:
Fix "mb" in the doc, add one more example to the "perf impact" section

EgorBo commented:
I'm merging the PR, we should see improvements at the PGO tab at https://aka.ms/aspnet/benchmarks in a couple of days.
Meanwhile I'll work on the queue of methods pending call-counting because this PR introduces new compilations hence delays promotion to tier1, I already have a couple of improvements in that area.

AndyAyersMS commented:
Keep a close eye on BDN results from the lab -- we may need to adjust things since some benchmarks implicitly rely on the old tiering strategy.

EgorBo commented:
![image](https://user-images.githubusercontent.com/523221/198364021-3cc4163a-7ae1-4aea-acc3-d9694e7fd9b7.png)

🙂
